File: src/harness/cursor/hooks.rs
//! Cursor hook rendering + support.
//!
//! Cursor's native matcher is coarser than Claude's (no `Glob`, no regex alternations like
//! `Edit|Write`, no `mcp__*` syntax). Rather than lossily down-translate, we register one blanket
//! entry per Cursor lifecycle event whose command invokes `agentpack hook-exec dispatch ...`; the
//! router reads Cursor's stdin (`tool_name`), normalizes it to candidate Claude tool names, and
//! fires the stored specs whose original Claude matcher matches.

use serde_json::{json, Map, Value};

use crate::error::Result;
use crate::harness::HarnessTarget;
use crate::hooks::capabilities::SupportLevel;
use crate::hooks::ir::{ClaudeEvent, ClaudeHandler, HookBundle, NormalizedHook};
use crate::hooks::paths::{hook_dispatch_command, specs_dispatch_root};
use crate::hooks::render::{
    build_exec_spec_file, check_support, push_diag, HookRenderer, RenderContext, RenderedHookFile,
    RenderedHookFileContents, RenderedHookOutput,
};

pub(super) struct CursorHookRenderer;

impl HookRenderer for CursorHookRenderer {
    fn target(&self) -> HarnessTarget {
        HarnessTarget::Cursor
    }

    fn render(&self, bundle: &HookBundle, ctx: &RenderContext<'_>) -> Result<RenderedHookOutput> {
        let mut output = RenderedHookOutput::default();
        let mut entries_per_event: std::collections::BTreeMap<&'static str, Vec<Value>> =
            std::collections::BTreeMap::new();

        for hook in &bundle.hooks {
            let Some(step) = mapped_cursor_step(hook.event) else {
                push_diag(
                    &mut output,
                    "omitted",
                    hook,
                    "Cursor has no equivalent lifecycle step",
                );
                continue;
            };
            if !check_support(
                HarnessTarget::Cursor,
                hook,
                &mut output,
                "routed via agentpack hook-exec dispatch",
                "routed via agentpack hook-exec dispatch",
            )? {
                continue;
            }
            // Always write the spec so the dispatcher can find it; the command line emitted
            // in hooks.json is blanket (one per event) so we deliberately ignore the return path.
            let _ =
                build_exec_spec_file(HarnessTarget::Cursor, hook, hook.event, ctx, &mut output)?;

            entries_per_event.entry(step).or_default();
            add_blanket_entry(
                &mut entries_per_event,
                step,
                hook.event,
                ctx,
                needs_fail_closed(hook),
            );
        }

        if entries_per_event.is_empty() {
            return Ok(output);
        }

        let mut hooks_map = Map::new();
        for (step, entries) in entries_per_event {
            hooks_map.insert(step.into(), Value::Array(entries));
        }

        output.files.push(RenderedHookFile {
            path: ctx.target_root.join("hooks/hooks.json"),
            contents: RenderedHookFileContents::Json(json!({
                "version": 1,
                "hooks": hooks_map,
            })),
        });
        Ok(output)
    }
}

/// Support level for emulating a Claude hook event+handler on Cursor.
pub(super) fn cursor_support(event: ClaudeEvent, handler: &ClaudeHandler) -> SupportLevel {
    match event {
        ClaudeEvent::Notification => SupportLevel::Unsupported {
            reason: "Cursor has no notification hook surface",
        },
        ClaudeEvent::PermissionRequest => match handler {
            ClaudeHandler::Http(_) | ClaudeHandler::Agent(_) => SupportLevel::Degraded {
                reason: "Cursor permission hooks are decomposed into preToolUse bridge commands",
            },
            _ => SupportLevel::Degraded {
                reason: "Cursor models permission requests as preToolUse instead of a dedicated event",
            },
        },
        ClaudeEvent::SessionStart | ClaudeEvent::PreCompact => match handler {
            ClaudeHandler::Http(_) | ClaudeHandler::Agent(_) => SupportLevel::Degraded {
                reason: "Cursor supports the lifecycle but requires bridge execution for this handler type",
            },
            _ => SupportLevel::Degraded {
                reason: "Cursor cannot preserve Claude trigger-specific matchers for this lifecycle event",
            },
        },
        _ => match handler {
            ClaudeHandler::Command(_) | ClaudeHandler::Prompt(_) => SupportLevel::Native,
            ClaudeHandler::Http(_) | ClaudeHandler::Agent(_) => SupportLevel::Emulated,
        },
    }
}

/// Insert a single dispatcher entry for `(cursor_step, claude_event)` if one isn't there yet.
/// Cursor fires every entry registered under a step; one per (step,event) is all we need.
fn add_blanket_entry(
    entries_per_event: &mut std::collections::BTreeMap<&'static str, Vec<Value>>,
    step: &'static str,
    event: ClaudeEvent,
    ctx: &RenderContext<'_>,
    fail_closed: bool,
) {
    let specs_dir = specs_dispatch_root(HarnessTarget::Cursor, ctx.target_root);
    let command = hook_dispatch_command(HarnessTarget::Cursor, event.as_claude_str(), &specs_dir);
    let entries = entries_per_event.entry(step).or_default();
    let already = entries.iter().any(|entry| {
        entry
            .get("command")
            .and_then(Value::as_str)
            .is_some_and(|c| c == command)
    });
    if already {
        return;
    }
    let mut obj = Map::new();
    obj.insert("type".into(), Value::String("command".into()));
    obj.insert("command".into(), Value::String(command));
    if fail_closed {
        obj.insert("failClosed".into(), Value::Bool(true));
    }
    entries.push(Value::Object(obj));
}

fn needs_fail_closed(hook: &NormalizedHook) -> bool {
    hook.is_strict() || matches!(hook.event, ClaudeEvent::PermissionRequest)
}

fn mapped_cursor_step(event: ClaudeEvent) -> Option<&'static str> {
    match event {
        ClaudeEvent::PreToolUse => Some("preToolUse"),
        ClaudeEvent::PostToolUse => Some("postToolUse"),
        ClaudeEvent::UserPromptSubmit => Some("beforeSubmitPrompt"),
        ClaudeEvent::Stop => Some("stop"),
        ClaudeEvent::SubagentStop => Some("subagentStop"),
        ClaudeEvent::SessionStart => Some("sessionStart"),
        ClaudeEvent::SessionEnd => Some("sessionEnd"),
        ClaudeEvent::PreCompact => Some("preCompact"),
        ClaudeEvent::PermissionRequest => Some("preToolUse"),
        ClaudeEvent::Notification => None,
    }
}

File: internal/harness/cursor/hooks.go
package cursor

import (
	"path/filepath"

	"github.com/OlegHQ/agentpack/internal/harness"
	"github.com/OlegHQ/agentpack/internal/hooks"
)

type HookRenderer struct{}

func (HookRenderer) Target() harness.Target { return harness.Cursor }

func (HookRenderer) Render(bundle hooks.Bundle, context hooks.RenderContext) (hooks.RenderOutput, error) {
	var output hooks.RenderOutput
	events := make(map[string][]any)
	seen := make(map[string]bool)
	for _, hook := range bundle.Hooks {
		step, mapped := cursorStep(hook.Event)
		if !mapped {
			hooks.PushDiagnostic(&output, "omitted", hook, "Cursor has no equivalent lifecycle step")
			continue
		}
		keep, err := hooks.CheckSupport(harness.Cursor, hook, Support(hook.Event, hook.Handler), &output, "routed via agentpack hook-exec dispatch", "routed via agentpack hook-exec dispatch")
		if err != nil {
			return hooks.RenderOutput{}, err
		}
		if !keep {
			continue
		}
		if _, err := hooks.BuildExecutionSpec(harness.Cursor, hook, hook.Event, context, &output); err != nil {
			return hooks.RenderOutput{}, err
		}
		key := step + "\x00" + string(hook.Event)
		if seen[key] {
			continue
		}
		seen[key] = true
		command := hooks.HookDispatchCommand(harness.Cursor, hook.Event, hooks.HookAssetRoot(harness.Cursor, context.TargetRoot))
		entry := map[string]any{"type": "command", "command": command}
		if hook.Strict() || hook.Event == hooks.PermissionRequest {
			entry["failClosed"] = true
		}
		events[step] = append(events[step], entry)
	}
	if len(events) == 0 {
		return output, nil
	}
	hookMap := make(map[string]any, len(events))
	for event, entries := range events {
		hookMap[event] = entries
	}
	output.Files = append(output.Files, hooks.RenderedFile{Path: filepath.Join(context.TargetRoot, "hooks", "hooks.json"), JSON: map[string]any{"version": 1, "hooks": hookMap}})
	return output, nil
}

func Support(event hooks.Event, handler hooks.Handler) hooks.Support {
	switch event {
	case hooks.Notification:
		return hooks.Support{Kind: hooks.Unsupported, Reason: "Cursor has no notification hook surface"}
	case hooks.PermissionRequest:
		if handler.Kind == hooks.HTTPHandler || handler.Kind == hooks.AgentHandler {
			return hooks.Support{Kind: hooks.Degraded, Reason: "Cursor permission hooks are decomposed into preToolUse bridge commands"}
		}
		return hooks.Support{Kind: hooks.Degraded, Reason: "Cursor models permission requests as preToolUse instead of a dedicated event"}
	case hooks.SessionStart, hooks.PreCompact:
		if handler.Kind == hooks.HTTPHandler || handler.Kind == hooks.AgentHandler {
			return hooks.Support{Kind: hooks.Degraded, Reason: "Cursor supports the lifecycle but requires bridge execution for this handler type"}
		}
		return hooks.Support{Kind: hooks.Degraded, Reason: "Cursor cannot preserve Claude trigger-specific matchers for this lifecycle event"}
	}
	if handler.Kind == hooks.CommandHandler || handler.Kind == hooks.PromptHandler {
		return hooks.Support{Kind: hooks.Native}
	}
	return hooks.Support{Kind: hooks.Emulated}
}

func cursorStep(event hooks.Event) (string, bool) {
	switch event {
	case hooks.PreToolUse, hooks.PermissionRequest:
		return "preToolUse", true
	case hooks.PostToolUse:
		return "postToolUse", true
	case hooks.UserPromptSubmit:
		return "beforeSubmitPrompt", true
	case hooks.Stop:
		return "stop", true
	case hooks.SubagentStop:
		return "subagentStop", true
	case hooks.SessionStart:
		return "sessionStart", true
	case hooks.SessionEnd:
		return "sessionEnd", true
	case hooks.PreCompact:
		return "preCompact", true
	default:
		return "", false
	}
}
